api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java

```diff
@@ -223,7 +223,7 @@ public Patient getPatientOrPromotePerson(Integer patientOrPersonId) {
 		try {
 			patient = Context.getPatientService().getPatient(patientOrPersonId);
 		}
-		catch (ClassCastException ex) {
```

Comment:
Just like the original exception, i would explicitly use `WrongClassException` instead of the generic `Exception`.

Secondly, even with the changes in this pull request, you would still get an `UnexpectedRollbackException` because of the above call `Context.getPatientService().getPatient(patientOrPersonId)` which marks the transaction for rollback. Therefore, you need to either change `Context.getPatientService().getPatient(patientOrPersonId)` to `getPatient(patientOrPersonId)` or add `noRollbackFor = WrongClassException.class` to the transaction definition for the `getPatient(Integer patientId)` method.

Reply:
@dkayiwa , thanks for that insight. I'd love to learn more about this from you.  Conceptually I'm confused about why any method annotated as a read-only transaction would ever result in an UnexpectedRollbackException - isn't the point of a read-only transaction that you want to prevent changes from getting written?  What is there to roll back?

From the perspective of our code, I'm not entirely sure why limiting our exception handling so narrowly will lead to better behavior.  Why not catch all Exceptions here and then attempt to load by person?  What type of error would we be catching that we feel we'd be better off throwing back to the calling code?